sw/device/tests/dif/dif_hmac_device_test.c

```diff
@@ -0,0 +1,290 @@
+ */
+static bool run_hmac(dif_hmac_t *hmac) {
+  switch (dif_hmac_run(hmac)) {
+    case kDifHmacOk:
+      break;
+    case kDifHmacBadArg:
+      LOG_ERROR("Invalid arguments encountered running HMAC.\r\n");
+      return false;
+    default:
+      LOG_ERROR("Unknown error encountered running HMAC.\r\n");
+      return false;
+  }
+
+  return true;
+}
+
+/**
+ * Spin until HMAC is done running and the digest is available. Return |true| if
+ * there are no errors.
+ */
+static bool wait_for_hmac_done(dif_hmac_t *hmac) {
+  bool hmac_done;
+  do {
+    switch (dif_hmac_done(hmac, &hmac_done)) {
+      case kDifHmacOk:
+        break;
+      case kDifHmacBadArg:
+        LOG_ERROR(
+            "Invalid arguments encountered checking HMAC run status.\r\n");
+        return false;
+      default:
+        LOG_ERROR("Unknown error encountered checking HMAC run status.\r\n");
+        return false;
+    }
+  } while (!hmac_done);
+
+  return true;
+}
+
+/**
+ * Read the HMAC digest and compare it to the expected result. Return |true| if
+ * there are no errors and the digests match.
+ */
+static bool check_digest(const dif_hmac_t *hmac,
+                         const uint32_t *expected_digest) {
+  dif_hmac_digest_t digest_result;
+  switch (dif_hmac_digest_read(hmac, &digest_result)) {
+    case kDifHmacOk:
+      break;
+    case kDifHmacBadArg:
+      LOG_ERROR("Invalid arguments encountered reading HMAC digest.\r\n");
+      return false;
+    default:
+      LOG_ERROR("Unknown error encountered reading HMAC digest.\r\n");
+      return false;
+  }
+
+  bool matches = true;
+  for (int i = 0; i < 8; ++i) {
+    if (expected_digest[i] != digest_result.digest[i]) {
+      LOG_ERROR(
+          "Digest mismatch, expected: 0x%X at index [%i] but got: "
+          "0x%X.\r\n",
+          expected_digest[i], i, digest_result.digest[i]);
+      matches = false;
+    }
+  }
+
+  return matches;
+}
+
+static bool run_test_impl(const dif_hmac_config_t *config, const char *data,
+                          size_t len, const uint32_t *expected_digest) {
+  dif_hmac_t hmac;
+  if (!test_setup(config, &hmac)) {
+    return false;
+  }
+
+  if (!push_message(&hmac, data, len)) {
+    return false;
+  }
+
+  if (!wait_for_fifo_empty(&hmac)) {
+    return false;
+  }
+
+  if (!check_message_length(&hmac, len * 8)) {
+    return false;
+  }
+
+  if (!run_hmac(&hmac)) {
+    return false;
+  }
+
+  if (!wait_for_hmac_done(&hmac)) {
+    return false;
+  }
+
+  if (!check_digest(&hmac, expected_digest)) {
+    return false;
+  }
+
+  return true;
+}
+
+void run_test(char *test_name, const dif_hmac_config_t *config,
+              const char *data, size_t len, const uint32_t *expected_digest) {
+  LOG_INFO("Running test \"");
+  uart_send_str(test_name);
+  uart_send_str("\":\r\n");
+  if (run_test_impl(config, data, len, expected_digest)) {
+    LOG_INFO("PASS!\r\n");
+  } else {
+    LOG_INFO("FAIL!\r\n");
+  }
+}
+
+int main(int argc, char **argv) {
```

Comment:
I believe Sri's `test_main` PR landed, so I would consider refactoring this to use that, instead.